test/gpu/native/studies/shoc/sort.chpl

```diff
@@ -0,0 +1,679 @@
+            writeln("Sort Parity ",parity, " N\n");
+        }
+        sortDb.addToDatabase("%{#####}".format(size), sortRate);
+        pciDb.addToDatabase("%{#####}".format(size), pciRate);
+        parityDb.addToDatabase("%{#####}".format(size), parity);
+    }
+    if(output) {
+        sortDb.printDatabaseStats();
+        pciDb.printDatabaseStats();
+        parityDb.printDatabaseStats();
+    }
+    if(perftest){
+        sortDb.printPerfStats();
+        pciDb.printPerfStats();
+        parityDb.printPerfStats();
+
+    }
+}
+
+startGPUDiagnostics();
+runSort();
+stopGPUDiagnostics();
+writeln(getGPUDiagnostics());
+
+
+proc radixSortStep(nbits: uint(32), startbit: uint(32),
+                    ref keys : [] uint(32), ref values : [] uint(32),
+                    ref tempKeys : [] uint(32), ref tempValues : [] uint(32),
+                    ref counters : [] uint(32), ref countersSum : [] uint(32),
+                    ref blockOffsets : [] uint(32),
+                    ref scanBlockSums : [] list(uint(32)), // scanBlockSums is an array of lists
+                    numElements: uint(32)){
+
+    // Threads have either 4 or 2 elements each
+    const radixGlobalWorkSize   = numElements / 4;
+    const findGlobalWorkSize    = numElements / 2;
+    const reorderGlobalWorkSize = numElements / 2;
+
+    // Radix Kernel uses block size of 128, others use 256
+    // (same as scan)
+    const radixBlocks   = radixGlobalWorkSize   / SORT_BLOCK_SIZE;
+    const findBlocks    = findGlobalWorkSize    / SCAN_BLOCK_SIZE;
+    const reorderBlocks = reorderGlobalWorkSize / SCAN_BLOCK_SIZE;
+
+    // radixSortBlocks Kernel here
+    radixSortBlocks(radixGlobalWorkSize, nbits, startbit, tempKeys, tempValues, keys, values);
+
+    // findRadixOffset Kernel here
+    // Use offsets again
+    findRadixOffsets(findGlobalWorkSize, tempKeys, counters, blockOffsets, startbit, numElements, findBlocks);
+
+    scanArrayRecursive(countersSum, counters, 16*reorderBlocks, 0, scanBlockSums);
+
+    // reorderData Kernel here
+    reorderData(reorderGlobalWorkSize, startbit, keys, values, tempKeys, tempValues,
+         blockOffsets, countersSum, counters, reorderBlocks);
+}
+
+proc scanArrayRecursive(ref outArray : [] uint(32), ref inArray : [] uint(32),
+                        numElements, level, ref blockSums : [] list(uint(32))){
+    // Kernels hadle 8 elements per thread
+    const numBlocks : uint(32) = max(1,ceil(numElements:real
+        / (4.0:real * SCAN_BLOCK_SIZE)) : uint(32)) : uint(32);
+
+    const sharedEltsPerBlock : uint(32) = (SCAN_BLOCK_SIZE * 2) : uint(32);
+    const sharedMemSize : uint(32) = numBytes(uint(32)) * sharedEltsPerBlock;
+
+    const fullBlock : bool = (numElements == numBlocks * 4 * SCAN_BLOCK_SIZE);
+
+    // CUDA version uses dim3 for no apparent reason because they
+    // created dim3 with only x populated for grid and block size
+
+    //execute scan
+    if(numBlocks>1){
```

Comment:
Do you know if you run this with all your problem sizes if we get code coverage across both branches of this `if/else`? Have you tested that this works with both code paths?

I'm curious if you know / have a theory for why both of these versions exist? Is the fullBlock version better for testing purposes or something?

Reply:
Both of these paths are taken in our code so yes we do have branch coverage for that `if/else`.
I do believe that this is not a performance thing, rather for correctness of how the CUDA version does copies of `uint4`s.